frontend/lib/src/components/elements/Popover/Popover.tsx

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { ReactElement } from "react"
+
+import { useTheme } from "@emotion/react"
+import { ExpandMore, ExpandLess } from "@emotion-icons/material-outlined"
+
+import { hasLightBackgroundColor } from "@streamlit/lib/src/theme"
+import { StyledIcon } from "@streamlit/lib/src/components/shared/Icon"
+import { Block as BlockProto } from "@streamlit/lib/src/proto"
+import BaseButton, {
+  BaseButtonKind,
+  BaseButtonSize,
+} from "@streamlit/lib/src/components/shared/BaseButton"
+import StreamlitMarkdown from "@streamlit/lib/src/components/shared/StreamlitMarkdown"
+import IsSidebarContext from "@streamlit/lib/src/components/core/IsSidebarContext"
+
+import { Popover as UIPopover, TRIGGER_TYPE, PLACEMENT } from "baseui/popover"
+
+import { StyledPopoverButtonIcon } from "./styled-components"
+
+export interface PopoverProps {
+  element: BlockProto.Popover
+  empty: boolean
+}
+
+const Popover: React.FC<PopoverProps> = ({
+  element,
+  empty,
+  children,
+}): ReactElement => {
+  const [open, setOpen] = React.useState(false)
+  const isInSidebar = React.useContext(IsSidebarContext)
+  const theme = useTheme()
+
+  const lightBackground = hasLightBackgroundColor(theme)
```

Comment:
Should this call be memoized and only be re-called when `theme` changes? It seems to call some library function under-the-hood, so I am not sure whether that call is expensive and we would need this.

Reply:
We tested the `hasLightBackgroundColor` performance some time ago, and it was negligible. But this is indeed another aspect we should just precompute as part of the theme. I might do this as a follow-up refactoring in the code base.

Reply:
I added it to tech debt [here](https://www.notion.so/snowflake-corp/Tech-Debt-DB-8613db6852c84a039c4e1c8e929507c4?p=09278d1c715f4e8ea5fc1a0019911ab2&pm=s).